File: node/parametric_processor.js
const Model = require('./gen/model_pb');
const Equation = Model.ParametricPattern.Equation;

function evaluateEquation(t, equation) {
  switch (equation.getEquationTypeCase()) {
    case Equation.EquationTypeCase.BINARY_EQUATION:
      return evaluateBinaryEquation(t, equation.getBinaryEquation());
    case Equation.EquationTypeCase.UNARY_EQUATION:
      return evaluateUnaryEquation(t, equation.getUnaryEquation());
    case Equation.EquationTypeCase.CONSTANT:
      return equation.getConstant();
    case Equation.EquationTypeCase.VARIABLE:
      return t;
  }
}

function evaluateBinaryEquation(t, binaryEquation) {
  const left = evaluateEquation(t, binaryEquation.getEquationLeft());
  const right = evaluateEquation(t, binaryEquation.getEquationRight());
  switch (binaryEquation.getOperation()) {
    case Equation.BinaryEquation.BinaryOperationType.ADD:
      return left + right;
    case Equation.BinaryEquation.BinaryOperationType.SUBTRACT:
      return left - right;
    case Equation.BinaryEquation.BinaryOperationType.MULTIPLY:
      return left * right;
    case Equation.BinaryEquation.BinaryOperationType.DIVIDE:
      return left / right;
    case Equation.BinaryEquation.BinaryOperationType.POWER:
      return Math.pow(left, right);
    case Equation.BinaryEquation.BinaryOperationType.ROOT:
      if (left == 0 || (right == 0 && left <= 0) || (right < 0 && left != Math.abs(left))) {
        return 0;
      }
      return pow(right, 1 / left);
    case Equation.BinaryEquation.BinaryOperationType.LOG:
      if (left <= 0 || right <= 0 || left == 1) {
        return 0;
      }
      return Math.log(right) / Math.log(left);
    case Equation.BinaryEquation.BinaryOperationType.ATAN2:
      return Math.atan2(left, right);
  }
}

function evaluateUnaryEquation(t, unaryEquation) {
  subValue = evaluateEquation(t, unaryEquation.getEquation());
  switch (unaryEquation.getOperation()) {
    case Equation.UnaryEquation.UnaryOperationType.SIN:
      return Math.sin(subValue);
    case Equation.UnaryEquation.UnaryOperationType.COS:
      return Math.cos(subValue);
    case Equation.UnaryEquation.UnaryOperationType.TAN:
      return Math.tan(subValue);
    case Equation.UnaryEquation.UnaryOperationType.ASIN:
      return Math.asin(subValue);
    case Equation.UnaryEquation.UnaryOperationType.ACOS:
      return Math.acos(subValue);
    case Equation.UnaryEquation.UnaryOperationType.ATAN:
      return Math.atan(subValue);
    case Equation.UnaryEquation.UnaryOperationType.NEGATIVE:
      return -subValue;
    case Equation.UnaryEquation.UnaryOperationType.STEP:
      return subValue >= 0 ? 1 : 0;
    case Equation.UnaryEquation.UnaryOperationType.SIGN:
      return subValue == 0 ? 0 : (subValue > 0 ? 1 : -1);
    case Equation.UnaryEquation.UnaryOperationType.ABS:
      return Math.abs(subValue);
  }
}

function getCoordinatesFromParametricFunction(pattern) {
  const result = [];
  for (let t = pattern.getStartT(); t <= pattern.getEndT(); t += pattern.getTStep()) {
    const xOrTheta = evaluateEquation(t, pattern.getEquationForXOrTheta());
    const yOrR = evaluateEquation(t, pattern.getEquationForYOrR());
    const coordinate = new Model.CartesianCoordinate();
    switch (pattern.getType()) {
      case Model.ParametricPattern.Type.CARTESIAN:
        coordinate.setX(xOrTheta);
        coordinate.setY(yOrR);
        break;
      case Model.ParametricPattern.Type.POLAR:
        coordinate.setX(yOrR * Math.cos(xOrTheta));
        coordinate.setY(yOrR * Math.sin(xOrTheta));
        break;
    }
    result.push(coordinate);
  }
  return result;
}

module.exports = {getCoordinatesFromParametricFunction};

File: src/parametric_processor.cc
#include "parametric_processor.h"

#include "model.pb.h"
#include <cmath>

namespace {

using sisyphus::ParametricPattern;
using sisyphus::ParametricPattern_Equation;
using sisyphus::ParametricPattern_Equation_UnaryEquation;
using sisyphus::ParametricPattern_Equation_BinaryEquation;

double EvaluateEquation(int t, const ParametricPattern_Equation& equation);
double EvaluateBinaryEquation(int t, const ParametricPattern_Equation_BinaryEquation& binary_equation);
double EvaluateUnaryEquation(int t, const ParametricPattern_Equation_UnaryEquation& unary_equation);

double EvaluateEquation(int t, const ParametricPattern_Equation& equation) {
  switch (equation.equation_type_case()) {
    case ParametricPattern_Equation::kBinaryEquation:
      return EvaluateBinaryEquation(t, equation.binary_equation());
    case ParametricPattern_Equation::kUnaryEquation:
      return EvaluateUnaryEquation(t, equation.unary_equation());
    case ParametricPattern_Equation::kConstant:
      return equation.constant();
    case ParametricPattern_Equation::kVariable:
      return t;
  }
}

double EvaluateBinaryEquation(int t, const ParametricPattern_Equation_BinaryEquation& binary_equation) {
  double left = EvaluateEquation(t, binary_equation.equation_left());
  double right = EvaluateEquation(t, binary_equation.equation_right());
  switch (binary_equation.operation()) {
    case ParametricPattern_Equation_BinaryEquation::ADD:
      return left + right;
    case ParametricPattern_Equation_BinaryEquation::SUBTRACT:
      return left - right;
    case ParametricPattern_Equation_BinaryEquation::MULTIPLY:
      return left * right;
    case ParametricPattern_Equation_BinaryEquation::DIVIDE:
      return left / right;
    case ParametricPattern_Equation_BinaryEquation::POWER:
      return pow(left, right);
    case ParametricPattern_Equation_BinaryEquation::ROOT:
      if (left == 0 || (right == 0 && left <= 0) || (right < 0 && left != std::abs(left))) {
        return 0;
      }
      return pow(right, 1 / left);
    case ParametricPattern_Equation_BinaryEquation::LOG:
      if (left <= 0 || right <= 0 || left == 1) {
        return 0;
      }
      return log(right) / log(left);
    case ParametricPattern_Equation_BinaryEquation::ATAN2:
      return atan2(left, right);
  }
}

double EvaluateUnaryEquation(int t, const ParametricPattern_Equation_UnaryEquation& unary_equation) {
  double sub_value = EvaluateEquation(t, unary_equation.equation());
  switch (unary_equation.operation()) {
    case ParametricPattern_Equation_UnaryEquation::SIN:
      return sin(sub_value);
    case ParametricPattern_Equation_UnaryEquation::COS:
      return cos(sub_value);
    case ParametricPattern_Equation_UnaryEquation::TAN:
      return tan(sub_value);
    case ParametricPattern_Equation_UnaryEquation::ASIN:
      return asin(sub_value);
    case ParametricPattern_Equation_UnaryEquation::ACOS:
      return acos(sub_value);
    case ParametricPattern_Equation_UnaryEquation::ATAN:
      return atan(sub_value);
    case ParametricPattern_Equation_UnaryEquation::NEGATIVE:
      return -sub_value;
    case ParametricPattern_Equation_UnaryEquation::STEP:
      return sub_value >= 0 ? 1 : 0;
    case ParametricPattern_Equation_UnaryEquation::SIGN:
      return sub_value == 0 ? 0 : (sub_value > 0 ? 1 : -1);
    case ParametricPattern_Equation_UnaryEquation::ABS:
      return std::abs(sub_value);
    case ParametricPattern_Equation_UnaryEquation::FLOOR:
      return floor(sub_value);
    case ParametricPattern_Equation_UnaryEquation::CEIL:
      return ceil(sub_value);
    case ParametricPattern_Equation_UnaryEquation::ROUND:
      return round(sub_value);
  }
}

}

std::vector<sisyphus::CartesianCoordinate>
ParametricProcessor::GetCoordinatesFromParametricFunction(
    const sisyphus::ParametricPattern& pattern) {
  std::vector<sisyphus::CartesianCoordinate> result;
  for (int t = pattern.start_t(); t <= pattern.end_t(); t += pattern.t_step()) {
    double x_or_theta = EvaluateEquation(t, pattern.equation_for_x_or_theta());
    double y_or_r = EvaluateEquation(t, pattern.equation_for_y_or_r());
    sisyphus::CartesianCoordinate coordinate;
    switch (pattern.type()) {
      case sisyphus::ParametricPattern::CARTESIAN:
        coordinate.set_x(x_or_theta);
        coordinate.set_y(y_or_r);
        break;
      case sisyphus::ParametricPattern::POLAR:
        coordinate.set_x(y_or_r * cos(x_or_theta));
        coordinate.set_y(y_or_r * sin(x_or_theta));
        break;
    }
    result.push_back(coordinate);
  }
  return result;
}
